### PR TITLE
make sure that modifications to scripts are applied to isolated engin…

### PR DESCRIPTION
**Pull request by objectundefined**
…es running on different threads with different nashorn contexts

### CODE REVIEW

**Comment by jexp** [comment]
Hey @objectundefined did you also have considered that we could/should add support for other scripting engines? e.g. by optionally passing in the other engine-name as params or by using different procedures?

**Comment by objectundefined** [comment]
I think the big hassle is exposing an appropriate api within each scripting engine. Like, it's one thing to pass everything into a groovy interpreter, it's another thing to write all the glue that makes it "gremlin". As far as Javascript goes, I think the few helper methods you've added definitely help, but I'd take a look at Cayley and the methods they add to the JS runtime to make it feel "gremlin-ish".

If this project were to grow into a really powerful utility, it would have to fill gaps that exist within cypher. That means 1) a traversal API for when pattern-matching doesn't cut it, 2) non-stream-oriented conditional logic, etc.

**tldr;** Not unless time is spent on idiomatic APIs per language/runtime.

**Comment by jexp** [comment]
Thanks a lot for your feedback.

Many of the things you mention can be done by just using APOC from cypher.

This is more to allow a scripting language user to quickly achieve sth that might not be possible in Cypher, like a more imperative operation.

I'm not sure if a gremlin-ish API the goal or a more javascript-ish.

**Comment by objectundefined** [comment]
I'm quite the happy APOC user as it stands. i never really had a particular need for a JS runtime in neo, as much as i thought it was a fun little project to hack on. It's always going to be a little bit funky moving between things like ResourceIterator and collection types in dynamic languages like JS. Maybe the best way of doing this is to shim GraphDatabaseService to normalize use of Iterators and Iterables, and java-idiomatic interfaces that don't translate well. I think that might be a better experience than a bag of helper functions.

**Comment by jexp** [comment]
Great idea, so that would include GDS, Node, Relationship, Result and the manual Indexes.

**Comment by jexp** [comment]
Thanks for the apoc praise
